### ISSUE TITLE
Regarding a line of codes

### ISSUE THREAD

**Issue by BNAadministrator3** [liuzechun/ReActNet#10]
Thank you for your terrfic work!
I've noticed:
binary_weights = binary_weights_no_grad.detach() - cliped_weights.detach() + cliped_weights
which appears at the line 83 in the ReActNet->mobilenet->1_step1->reactnet.py
I wonder what's the line of codes aims for?

**Comment by theqihao** [comment]
when forward, use the value of binary_weights_no_grad. when backword, use the cliped_weights's derivative.

**Comment by BNAadministrator3** [comment]
Thanks!

**Comment by BNAadministrator3** [comment]
By the way,
I noticed in each folder, no matter 'mobilenet' or 'resnet', there are two subfolders that are named '1_step1' and '2_step2' respectively.  It seems the contents of '1_step1' and '2_step2' are nearly the same. What is the difference b.t. the contents of '1_step1' and '2_step2'?

**Comment by feizi** [comment]
(1) Step1: binarizing activations

(2) Step2: binarizing weights + activations

**Comment by liuzechun** [comment]
Hi @BNAadministrator3 thank you for your nice words and your interest in our work! Sorry for my late reply. 
I think @feizi and @theqihao already answered the question. Big thanks to @feizi and @theqihao! If you have any further questions, just let me know!

**Comment by rrryan2016** [comment]
Guess similar intention in `BinaryActivation`, as,
 
https://github.com/liuzechun/ReActNet/blob/09ab2ea90162e9d7eb32cfded68e8428d3a30ea4/mobilenet/1_step1/reactnet.py#L42-L45

But any explanation for why calculate `out3` in this way?

**Comment by liuzechun** [comment]
Because out3 is a piece-wise linear function.

> Guess similar intention in `BinaryActivation`, as,
> 
> https://github.com/liuzechun/ReActNet/blob/09ab2ea90162e9d7eb32cfded68e8428d3a30ea4/mobilenet/1_step1/reactnet.py#L42-L45
> 
> But any explanation for why calculate `out3` in this way?
